### PR TITLE
Fix issue #584 add option to set AI daily cap and hit daily cap stop notify

### LINKED ISSUE #584: Feature Request: Support for Application Insights billing & pricing
### Terraform Version
Terraform v0.11.0
+ provider.azurevm v0.3.3

### Affected Resource(s)
- azurerm_application_insights

### Expected Behavior
It would be great if we could create a "microsoft.insights/components/CurrentBillingFeatures" resource so we can set billing features, data caps etc.

At present we need to work around this by configuring via an "azurerm_template_deployment" resource.

### PR DESCRIPTION
**Pull request by NickMetz**
This PR should fix #584, so we can set now a daily cap for Applications Insight in GB. Beside setting a daily cap for Applications Insight it's also possible now to disable notifications if daily cap is hit. 

Compared to other resources this configuration is separated to it's own billing feature api. Which requires to update billing feature api once Application Insight was created.

### CODE REVIEW

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 1]
```diff
@@ -167,13 +184,32 @@ func resourceArmApplicationInsightsCreateUpdate(d *schema.ResourceData, meta int
...
+	applicationInsightsComponentBillingFeatures := insights.ApplicationInsightsComponentBillingFeatures{
+		CurrentBillingFeatures: billingFeatureRead.CurrentBillingFeatures,
+		DataVolumeCap: &insights.ApplicationInsightsComponentDataVolumeCap{
+			Cap:                            dailyCap,
+			StopSendNotificationWhenHitCap: stopSendNotificationWhenHitCap,
+		},
+	}
+
+	_, err = billingFeatureClient.Update(ctx, resGroup, name, applicationInsightsComponentBillingFeatures)
+	if err != nil {
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: These lines can be merged

File: azurerm/internal/services/applicationinsights/client/client.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ type Client struct {
 	APIKeysClient        *insights.APIKeysClient
 	ComponentsClient     *insights.ComponentsClient
 	WebTestsClient       *insights.WebTestsClient
+	BillingFeatureClient *insights.ComponentCurrentBillingFeaturesClient
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/client/client.go]: Could we just call this
```suggestion
	BillingClient *insights.ComponentCurrentBillingFeaturesClient
```

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 3]
```diff
@@ -89,6 +89,19 @@ func resourceArmApplicationInsights() *schema.Resource {
...
+				Type:         schema.TypeFloat,
+				Optional:     true,
+				Default:      100,
+				ValidateFunc: validation.FloatBetween(0, 1000),
+			},
+
+			"stop_send_notification_when_hit_cap": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Could we make this as computed and get the default from the api?

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 4]
```diff
@@ -89,6 +89,19 @@ func resourceArmApplicationInsights() *schema.Resource {
 
 			"tags": tags.Schema(),
 
+			"daily_cap": {
+				Type:         schema.TypeFloat,
+				Optional:     true,
+				Default:      100,
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Could we make this as computed and get the default from the api?

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 5]
```diff
@@ -105,7 +118,9 @@ func resourceArmApplicationInsights() *schema.Resource {
 
 func resourceArmApplicationInsightsCreateUpdate(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*clients.Client).AppInsights.ComponentsClient
+	billingFeatureClient := meta.(*clients.Client).AppInsights.BillingFeatureClient
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]:
```suggestion
	billingClient := meta.(*clients.Client).AppInsights.BillingFeatureClient
```

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 6]
```diff
@@ -167,13 +184,32 @@ func resourceArmApplicationInsightsCreateUpdate(d *schema.ResourceData, meta int
...
+	}
+
 	d.SetId(*read.ID)
 
 	return resourceArmApplicationInsightsRead(d, meta)
 }
 
 func resourceArmApplicationInsightsRead(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*clients.Client).AppInsights.ComponentsClient
+	billingFeatureClient := meta.(*clients.Client).AppInsights.BillingFeatureClient
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]:
```suggestion
	billingClient := meta.(*clients.Client).AppInsights.BillingFeatureClient
```

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 7]
```diff
@@ -196,6 +232,11 @@ func resourceArmApplicationInsightsRead(d *schema.ResourceData, meta interface{}
 		return fmt.Errorf("Error making Read request on AzureRM Application Insights '%s': %+v", name, err)
 	}
 
+	billingFeatureResp, err := billingFeatureClient.Get(ctx, resGroup, name)
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]:
```suggestion
	billingResp, err := billingFeatureClient.Get(ctx, resGroup, name)
```

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 8]
```diff
@@ -89,6 +89,19 @@ func resourceArmApplicationInsights() *schema.Resource {
 
 			"tags": tags.Schema(),
 
+			"daily_cap": {
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: This might be better as 
```suggestion
			"daily_traffic_cap_in_gb": {
```

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go [hunk 9]
```diff
@@ -89,6 +89,19 @@ func resourceArmApplicationInsights() *schema.Resource {
...
+				Type:         schema.TypeFloat,
+				Optional:     true,
+				Default:      100,
+				ValidateFunc: validation.FloatBetween(0, 1000),
+			},
+
+			"stop_send_notification_when_hit_cap": {
```
Comment (katbyte) [inline azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: This property is awfully odd, could we invert and reword it to
```suggestion
			"daily_traffic_cap_notifications_enabled": {
```

where true -> sends false, false -> send true

Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Hi @katbyte,
agree with you awfully odd. But i think we should name it the other way. Because in general people will only touch this option if they want to stop notifications if cap is hit e.g. disabled = true.

**daily_data_cap_notifications_disabled**
true -> don't send messages
false -> send messages
default is false

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 8 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: This option is related to data which is actually stored in applications insight. So my recommendation would be **daily_data_cap_in_gb**. What you think?

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 1 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Merged if condition and resource update.

File: azurerm/internal/services/applicationinsights/client/client.go (hunk 2 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/client/client.go]: Renamed from BillingFeatureClient to BillingClient.

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 5 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Renamed from billingFeatureClient to billingClient.

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 3 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Removed default switched to computed.

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 4 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Removed default switched to computed.

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 6 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Variables renamed.

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 7 above)
Comment (NickMetz) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: Variables renamed.

**Comment by NickMetz** [comment]
@katbyte can you please check my two open questions. If my recommendation is ok, this PR is ready to merge.

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 8 above)
Comment (katbyte) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: works for me!

File: azurerm/internal/services/applicationinsights/resource_arm_application_insights.go (hunk 9 above)
Comment (katbyte) [reply azurerm/internal/services/applicationinsights/resource_arm_application_insights.go]: I'm ok with `daily_data_cap_notifications_disabled`, thanks!
